### PR TITLE
Clean up the concept of "pseudo ops"

### PR DESCRIPTION
**Pull request by tfoleyNV**
Built-in functions in the Slang standard library can be marked with `__intrinsic_op(...)` to indicate that they should not lower to functions in the IR, and that instead call sites to those functions should be translated directly to the IR.
There are two cases where `__intrinsic_op(...)` gets used:

1. In the case where the argument to `__intrinsic_op(...)` is an actual IR instruction opcode, the IR lowering logic directly translates a call into an instruction with the given opcode. The arguments to the call become the operands of the instruction.

2. In the case where the argument to `__intrinsic_op(...)` is one of a set of "pseudo" instruction opcodes, the IR lowering logic directly handles the lowering to IR with dedicated code. The operands to the call might be handled differently depending on the kind of operation.

The compound operators like `+=` are the most important example of these "pseudo" instructions.
It doesn't make sense to handle them as true function calls (although that would work semantically), nor does it make sense to have a single IR instruction with such complicated semantics.

An earlier version of the compiler used the same enumeration for both the true IR instruction opcodes and these "pseudo" opcodes, with the simple constraint that the pseudo opcodes were all negative while the real opcodes were positive. That design got changed up over a few refactorings, and because there was never a good explanation in the code itself of what "pseudo" opcodes were, we eventually ended up in a place where the in-memory and serialized IR encodings included logic to try to deal with the possibility of these "pseudo" opcodes, even though the entire design of the lowering pass meant that they'd never appear in generated IR.

This change tries to clean up the mess in a few ways:

* The terminology is now that these are "compound" intrinsic ops, to differentiate them from the more common case of intrinsic ops that map one-to-one to IR instructions.

* The declaration of the compound intrinsic ops is no longer in a file related to the IR, and doesn't use the `IR` naming prefix, so somebody looking at the IR opcodes cannot become confused and think the compound ops are allowed there.

* The IR encoding in memory and when serialized is updated to not account for or worry about the possibility of "pseudo" ops.

* The compound ops are declared in such a way that ensures their enumerant values are all negative, so that they are yet again trivially disjoint from the true IR opcodes.

A more drastic change might have split `__intrinsic_op` into two different modifier types: one for the trivial single-instruction case and one for the compound case.
Doing this would make the change more invasive, though, because there are places in the meta-code that generates the standard library that intentionally handle both single-instruction and compound ops (because built-in operators can translate to either case).

### CODE REVIEW

File: source/slang/slang-modifier-defs.h [hunk 1]
```diff
@@ -28,16 +28,26 @@ SIMPLE_MODIFIER(GloballyCoherent)
 #undef SIMPLE_MODIFIER
 
 // A modifier that marks something as an operation that
+// translates to special
```
Comment (jsmall-nvidia) [inline source/slang/slang-modifier-defs.h]: special...?

File: source/slang/slang-modifier-defs.h [hunk 2]
```diff
@@ -28,16 +28,26 @@ SIMPLE_MODIFIER(GloballyCoherent)
...
+    //
+    // If greather than or equal to zero, then `op`
+    // is an `IROp` and directly identifies an IR
+    // instruction that the intrinsic should translate to.
+    //
+    // If less than zero, then `op` is a `CompoundIntrinsicOp`
+    // which maps to zero or more IR instructions using
+    // special-case logic in the IR lowering phase.
+    //
+    FIELD_INIT(int32_t, op, 0)
```
Comment (jsmall-nvidia) [inline source/slang/slang-modifier-defs.h]: Perhaps would be preferable to have a typedef naming the Op than just the in32_t. Could also use an empty enum class, and a couple of free functions to determine what it is, and extract from (with asserts on invalid extractions).

File: source/slang/slang-modifier-defs.h (hunk 1 above)
Comment (tfoleyNV) [reply source/slang/slang-modifier-defs.h]: Garbage text due to copy-pasting that I meant to remove. Will clean up.

File: source/slang/slang-modifier-defs.h (hunk 2 above)
Comment (tfoleyNV) [reply source/slang/slang-modifier-defs.h]: I might do the `typedef` - that was something I thought about when writing the code but didn't end up doing. Actually adding strong semantics and so forth seems like overkill given just how few points in the code need to worry about this.

**Pull request merged**